File: Cpp/Algorithm/src/Chapter4/IsValid_20.cpp
#include "chapter4.h"

bool myAlgo::IsValid_20::isValid(string s)
{
	int n = s.size();
	if (n % 2 == 1) {
		return false;
	}
	unordered_map<char, char> map = {
			{')', '('},
			{']', '['},
			{'}', '{'}
	};

	stack<char> stack;
	for (char c : s) {
		// Óöµ½ÁËÓÒ±ßµÄÀ¨ºÅ
		if (map.count(c))
		{
			// top->peek
			if (stack.empty() || stack.top()!=map[c])
			{
				return false;
			}
			stack.pop();
		}
		else
		{
			stack.push(c);
		}
	}
	// "("
	return stack.empty();
}

File: JavaAlgorithm/src/main/java/algo/tzashinorpu/SpecialSubject/Stack_Queue/isValid_20.java
package algo.tzashinorpu.SpecialSubject.Stack_Queue;

import java.util.Stack;

public class isValid_20 {
    public boolean isValid(String s) {
        Stack<Character> stack = new Stack<>();
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (c == '(' || c == '[' || c == '{') {
                stack.push(c);
            } else {
                if (!stack.empty() && stack.peek() == pairOf(c)) {
                    stack.pop();
                } else return false;
            }
        }
        return stack.empty();
    }

    private Character pairOf(char c) {
        if (c == ')') return '(';
        if (c == ']') return '[';
        return '{';
    }
}
